Hey, thanks for the feedback. I've implemented most of your changes, but I'm struggling to get the server-side async working. Here's what the code looks like currently:

    ReceivedPacket::RpcExecute(rpc) => {
        let state = state.lock().unwrap();
        let client = Box::pin(Box::new(state.clients.get(&addr).unwrap().clone()));

        match state.rpc_actions.get(&rpc.entry_id) {
            Some(func) => {
                let func = Box::pin(Box::new(func.clone()));
                thread::spawn(move || {
                    Pin::into_inner(client)
                        .unbounded_send(Box::new(RpcResponse::new(
                            rpc.entry_id,
                            rpc.unique_id,
                            func(rpc.parameter),
                        )))
                        .unwrap();
                });
            }
            None => {
                Pin::into_inner(client)
                    .unbounded_send(Box::new(RpcResponse::new(
                        rpc.entry_id,
                        rpc.unique_id,
                        Vec::new(),
                    )))
                    .unwrap();
            }
        }
    }

And the error:

        Checking nt v1.0.3 (C:\Dev\jlmcmchl\nt-rs)
    error[E0597]: `state` does not live long enough
      --> src\proto\server\conn.rs:269:33
        |
    269 |                     let state = state.lock().unwrap();
        |                                 ^^^^^-------
        |                                 |
        |                                 borrowed value does not live long enough
        |                                 argument requires that `state` is borrowed for `'static`
        |
    269 |                     let state = state.lock().unwrap();
        |                                 ^^^^^-------
        |                                 |
        |                                 borrowed value does not live long enough
        |                                 argument requires that `state` is borrowed for `'static`
    ...
    311 | }
        | - `state` dropped here while still borrowed

    error[E0597]: `state` does not live long enough
      --> src\proto\server\conn.rs:272:27
        |
    272 |                     match state.rpc_actions.get(&rpc.entry_id) {
        |                           ^^^^^-------------------------------
        |                           borrowed value does not live long enough
        |                           argument requires that `state` is borrowed for `'static`
    ...
    295 |                 }
        |                 - `state` dropped here while still borrowed

    error: aborting due to 2 previous errors

    For more information about this error, try `rustc --explain E0597`.
    error: could not compile `nt`.

    To learn more, run the command again with --verbose.

I can't figure out how to solve the ownership issue here. Any ideas?

--- a/src/nt.rs
+++ b/src/nt.rs
@@ -153,7 +153,7 @@ impl NetworkTables<Server> {
     pub fn create_rpc(
         &mut self,
         data: EntryData,
-        callback: impl Fn(Vec<u8>) -> Vec<u8> + Send + RefUnwindSafe + 'static,
+        callback: impl Fn(Vec<u8>) -> Vec<u8> + Send + Sync + RefUnwindSafe + 'static,
     ) {
         self.state.lock().unwrap().create_rpc(data, callback);
     }

--- a/src/nt/callback.rs
+++ b/src/nt/callback.rs
@@ -19,6 +19,6 @@ pub type ConnectionAction = dyn FnMut(&SocketAddr) + Send + 'static;
 
 pub type Action = dyn FnMut(&EntryData) + Send + 'static;
 
-pub type RpcAction = dyn Fn(Vec<u8>) -> Vec<u8> + Send + RefUnwindSafe + 'static;
+pub type RpcAction = dyn Fn(Vec<u8>) -> Vec<u8> + Send + Sync + RefUnwindSafe + 'static;
 
 pub type RpcCallback = dyn Fn(Vec<u8>) + Send + 'static;

--- a/src/proto/client.rs
+++ b/src/proto/client.rs
@@ -27,7 +27,7 @@ pub struct ClientState {
     pub(crate) pending_entries: HashMap<String, Sender<u16>>,
     pub(crate) packet_tx: UnboundedSender<Box<dyn Packet>>,
     rpc_callbacks: HashMap<u16, Box<RpcCallback>>,
-    next_id: u16,
+    next_rpc_id: u16,
 }
 
 impl ClientState {
@@ -44,7 +44,7 @@ impl ClientState {
             pending_entries: HashMap::new(),
             packet_tx,
             rpc_callbacks: HashMap::new(),
-            next_id: 0,
+            next_rpc_id: 0,
         }));
 
         let rt_state = state.clone();
@@ -76,7 +76,7 @@ impl ClientState {
             pending_entries: HashMap::new(),
             packet_tx,
             rpc_callbacks: HashMap::new(),
-            next_id: 0,
+            next_rpc_id: 0,
         }));
 
         let rt_state = state.clone();
@@ -105,12 +105,13 @@ impl ClientState {
         parameter: Vec<u8>,
         callback: impl Fn(Vec<u8>) + Send + 'static,
     ) {
-        self.rpc_callbacks.insert(self.next_id, Box::new(callback));
+        self.rpc_callbacks
+            .insert(self.next_rpc_id, Box::new(callback));
         self.packet_tx
-            .unbounded_send(Box::new(RpcExecute::new(id, self.next_id, parameter)))
+            .unbounded_send(Box::new(RpcExecute::new(id, self.next_rpc_id, parameter)))
             .unwrap();
 
-        self.next_id += 1;
+        self.next_rpc_id += 1;
     }
 }
 

--- a/src/proto/client/conn.rs
+++ b/src/proto/client/conn.rs
@@ -265,7 +265,9 @@ fn handle_packet(packet: ReceivedPacket, state: &Arc<Mutex<ClientState>>) -> cra
         ReceivedPacket::RpcResponse(rpc) => {
             let mut state = state.lock().unwrap();
             if let Some(callback) = state.rpc_callbacks.remove(&rpc.unique_id) {
-                callback(rpc.result);
+                tokio::spawn(async move {
+                    callback(rpc.result);
+                });
             }
         }
         _ => {}

--- a/src/proto/server.rs
+++ b/src/proto/server.rs
@@ -10,8 +10,7 @@ use nt_network::{
 };
 use std::collections::HashMap;
 use std::net::SocketAddr;
-use std::panic;
-use std::panic::RefUnwindSafe;
+use std::panic::{self, RefUnwindSafe};
 
 use std::sync::{Arc, Mutex};
 use std::thread;
@@ -66,7 +65,7 @@ impl ServerState {
     pub fn create_rpc(
         &mut self,
         data: EntryData,
-        callback: impl Fn(Vec<u8>) -> Vec<u8> + Send + RefUnwindSafe + 'static,
+        callback: impl Fn(Vec<u8>) -> Vec<u8> + Send + Sync + RefUnwindSafe + 'static,
     ) {
         let id = self.create_entry(data).try_next().unwrap().unwrap();
         self.rpc_actions.insert(id, Box::new(callback));

--- a/src/proto/server.rs
+++ b/src/proto/server.rs
@@ -25,7 +25,7 @@ pub struct ServerState {
     callbacks: MultiMap<CallbackType, Box<Action>>,
     server_callbacks: MultiMap<ConnectionCallbackType, Box<ConnectionAction>>,
     next_id: u16,
-    rpc_actions: HashMap<u16, Box<RpcAction>>,
+    rpc_actions: HashMap<u16, Arc<RpcAction>>,
 }
 
 fn spawn_rt(ip: String, state: Arc<Mutex<ServerState>>, close_rx: Receiver<()>) {
@@ -68,7 +68,7 @@ impl ServerState {
         callback: impl Fn(Vec<u8>) -> Vec<u8> + Send + Sync + RefUnwindSafe + 'static,
     ) {
         let id = self.create_entry(data).try_next().unwrap().unwrap();
-        self.rpc_actions.insert(id, Box::new(callback));
+        self.rpc_actions.insert(id, Arc::new(callback));
     }
 
     pub fn call_rpc(&self, id: u16, parameter: Vec<u8>) -> Vec<u8> {

--- a/src/proto/server/conn.rs
+++ b/src/proto/server/conn.rs
@@ -11,8 +11,8 @@ use nt_network::{
     ServerHello, ServerHelloComplete,
 };
 use std::net::SocketAddr;
+use std::panic;
 use std::sync::{Arc, Mutex};
-use std::thread;
 use tokio::net::{TcpListener, TcpStream};
 use tokio_util::codec::Decoder;
 
@@ -268,25 +268,34 @@ where
                     let state = state.lock().unwrap();
                     let client = state.clients.get(&addr).unwrap().clone();
 
-                    if let Some(func) = state.rpc_actions.get(&rpc.entry_id) {
-                        let func = func.clone();
-                        thread::spawn(move || {
+                    match state.rpc_actions.get(&rpc.entry_id) {
+                        Some(func) => {
+                            let func = func.clone();
+                            tokio::spawn(async move {
+                                let result =
+                                    match panic::catch_unwind(|| func(rpc.parameter.clone())) {
+                                        Ok(res) => res,
+                                        Err(_) => Vec::new(),
+                                    };
+
+                                client
+                                    .unbounded_send(Box::new(RpcResponse::new(
+                                        rpc.entry_id,
+                                        rpc.unique_id,
+                                        result,
+                                    )))
+                                    .unwrap();
+                            });
+                        }
+                        None => {
                             client
                                 .unbounded_send(Box::new(RpcResponse::new(
                                     rpc.entry_id,
                                     rpc.unique_id,
-                                    func(rpc.parameter),
+                                    Vec::new(),
                                 )))
                                 .unwrap();
-                        });
-                    } else {
-                        client
-                            .unbounded_send(Box::new(RpcResponse::new(
-                                rpc.entry_id,
-                                rpc.unique_id,
-                                Vec::new(),
-                            )))
-                            .unwrap();
+                        }
                     }
                 }
                 _ => {}

--- a/src/proto/server/conn.rs
+++ b/src/proto/server/conn.rs
@@ -12,6 +12,7 @@ use nt_network::{
 };
 use std::net::SocketAddr;
 use std::sync::{Arc, Mutex};
+use std::thread;
 use tokio::net::{TcpListener, TcpStream};
 use tokio_util::codec::Decoder;
 
@@ -265,16 +266,28 @@ where
                 }
                 ReceivedPacket::RpcExecute(rpc) => {
                     let state = state.lock().unwrap();
-                    let result = state.call_rpc(rpc.entry_id, rpc.parameter.clone());
+                    let client = state.clients.get(&addr).unwrap().clone();
 
-                    state.clients.iter().for_each(|(_, tx)| {
-                        tx.unbounded_send(Box::new(RpcResponse::new(
-                            rpc.entry_id,
-                            rpc.unique_id,
-                            result.clone(),
-                        )))
-                        .unwrap()
-                    });
+                    if let Some(func) = state.rpc_actions.get(&rpc.entry_id) {
+                        let func = func.clone();
+                        thread::spawn(move || {
+                            client
+                                .unbounded_send(Box::new(RpcResponse::new(
+                                    rpc.entry_id,
+                                    rpc.unique_id,
+                                    func(rpc.parameter),
+                                )))
+                                .unwrap();
+                        });
+                    } else {
+                        client
+                            .unbounded_send(Box::new(RpcResponse::new(
+                                rpc.entry_id,
+                                rpc.unique_id,
+                                Vec::new(),
+                            )))
+                            .unwrap();
+                    }
                 }
                 _ => {}
             }

